pkg/webhook/authentication/http.go

```diff
@@ -0,0 +1,152 @@
+		wh.log.Error(err, "unable to process a request with an unknown content type", "content type", contentType)
+		reviewResponse = Errored(err)
+		wh.writeResponse(w, reviewResponse)
+		return
+	}
+
+	// Both v1 and v1beta1 TokenReview types are exactly the same, so the v1beta1 type can
+	// be decoded into the v1 type. The v1beta1 api is deprecated as of 1.19 and will be
+	// removed in v1.22. However the runtime codec's decoder guesses which type to
+	// decode into by type name if an Object's TypeMeta isn't set. By setting TypeMeta of an
+	// unregistered type to the v1 GVK, the decoder will coerce a v1beta1 TokenReview to v1.
+	// The actual TokenReview GVK will be used to write a typed response in case the
+	// webhook config permits multiple versions, otherwise this response will fail.
+	req := Request{}
+	ar := unversionedTokenReview{}
+	// avoid an extra copy
+	ar.TokenReview = &req.TokenReview
+	ar.SetGroupVersionKind(v1.SchemeGroupVersion.WithKind("TokenReview"))
+	_, actualTokRevGVK, err := authenticationCodecs.UniversalDeserializer().Decode(body, nil, &ar)
+	if err != nil {
+		wh.log.Error(err, "unable to decode the request")
+		reviewResponse = Errored(err)
+		wh.writeResponse(w, reviewResponse)
+		return
+	}
+	wh.log.V(1).Info("received request", "UID", req.UID, "kind", req.Kind)
+
+	if req.Spec.Token == "" {
+		err = errors.New("token is empty")
+		wh.log.Error(err, "bad request")
+		reviewResponse = Errored(err)
+		wh.writeResponse(w, reviewResponse)
+		return
+	}
+
+	// TODO: add panic-recovery for Handle
+	reviewResponse = wh.Handle(ctx, req)
+	wh.writeResponseTyped(w, reviewResponse, actualTokRevGVK)
+}
+
+// writeResponse writes response to w generically, i.e. without encoding GVK information.
+func (wh *Webhook) writeResponse(w io.Writer, response Response) {
+	wh.writeTokenResponse(w, response.TokenReview)
+}
+
+// writeResponseTyped writes response to w with GVK set to tokRevGVK, which is necessary
+// if multiple TokenReview versions are permitted by the webhook.
+func (wh *Webhook) writeResponseTyped(w io.Writer, response Response, tokRevGVK *schema.GroupVersionKind) {
+	ar := response.TokenReview
+
+	// Default to a v1 TokenReview, otherwise the API server may not recognize the request
+	// if multiple TokenReview versions are permitted by the webhook config.
+	if tokRevGVK == nil || *tokRevGVK == (schema.GroupVersionKind{}) {
+		ar.SetGroupVersionKind(v1.SchemeGroupVersion.WithKind("TokenReview"))
+	} else {
+		ar.SetGroupVersionKind(*tokRevGVK)
+	}
+	wh.writeTokenResponse(w, ar)
+}
+
+// writeTokenResponse writes ar to w.
+func (wh *Webhook) writeTokenResponse(w io.Writer, ar v1.TokenReview) {
+	err := json.NewEncoder(w).Encode(ar)
+	if err != nil {
+		wh.log.Error(err, "unable to encode the response")
+		wh.writeResponse(w, Errored(err))
+	} else {
+		res := ar
+		if log := wh.log; log.V(1).Enabled() {
+			log.V(1).Info("wrote response", "UID", res.UID, "authenticated", res.Status.Authenticated)
+		}
+	}
```

Comment:
```suggestion
        return
	}
	res := ar
	if log := wh.log; log.V(1).Enabled() {
		log.V(1).Info("wrote response", "UID", res.UID, "authenticated", res.Status.Authenticated)
	}
```

We can avoid unnecessary branching via early returns. See https://blog.codinghorror.com/flattening-arrow-code/ for some (exaggerated) examples on how to avoid nesting

Reply:
Sounds good, updating. We probably need to do a clean up of the admission package too, this again :p was a copy/pasta from there just updating the types used. Realistically it feels like there should be a way to abstract all these packages into fewer copied funcs and just passing around `ctrl.Object` and letting the handler convert to the proper type but I don't know if that would be necessary for this PR first.